### ISSUE TITLE
Support for calling `.toString()` on functions

### ISSUE THREAD

**Issue by anandthakker** [evanw/esbuild#2159]
Quoting the docs for reference: 

> It's somewhat common to call toString() on a JavaScript function object and then pass that string to some form of eval to get a new function object. This effectively "rips" the function out of the containing file and breaks links with all variables in that file. Doing this with esbuild is not supported and may not work. In particular, esbuild often uses helper methods to implement certain features and it assumes that JavaScript scope rules have not been tampered with.

(Examples of features that use helpers: `**`, `async` from JS; `keepNames` from esbuild.)

Do you see this as basically a hard constraint, or is there any chance that esbuild could support this under certain conditions?  Motivated here a scenario where I'm bundling an API with a third-party dependency that uses this technique.

One workaround that I think would work today is loading that dependency as a string that's evaluated at runtime, but I'm worried that'll affect performance.  Another is to leave that dependency external, patch it to export a global variable, and access it that way--but that's not ideal, either.

A sort of far-fetched idea on how _maaaybe_ it could be possible to bundle code like this with esbuild: have some way of marking a function to indicate that compilers must not introduce references outside of its scope (e.g. an  `/*#__ISOLATED__*/` annotation?).  Any such function would have to have local copies of any generated helpers inlined into it, which I'm guessing wouldn't be a deal breaker for things like `**` but maybe could be an issue for stuff like the generated module initializer functions that I've seen in esbuilt output.  But the latter could be prevented by a requirement that any external references in the function must resolve to the global scope (e.g. `Math` is fine, module-local names aren't).

**Comment by evanw** [comment]
> Do you see this as basically a hard constraint

Yes, basically. All of esbuild is designed with the assumption that scope lookup works the way it works in JavaScript. Messing with that fundamental assumption would likely involve rewriting large parts of the code. It would be a lot of maintenance overhead for me as it complicates everything, and is not something I'm interested in supporting.
